azurerm/data_source_resource.go

```diff
@@ -0,0 +1,181 @@
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"resource_id"},
+			},
+			"type": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"resource_id"},
+			},
+
+			"required_tags": tags.Schema(),
+
+			"resources": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"location": azure.SchemaLocationForDataSource(),
+						"tags":     tags.SchemaDataSource(),
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceArmResourceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).Resource.ResourcesClient
+	ctx, cancel := timeouts.ForRead(meta.(*ArmClient).StopContext, d)
+	defer cancel()
+
+	resourceGroupName := d.Get("resource_group_name").(string)
+	resourceName := d.Get("name").(string)
+	resourceType := d.Get("type").(string)
+	resourceID := d.Get("resource_id").(string)
+
+	var filter string
+
+	if resourceGroupName != "" {
+		v := fmt.Sprintf("resourceGroup eq '%s'", resourceGroupName)
+		filter = filter + v
+	}
+
+	if resourceName != "" {
+		if strings.Contains(filter, "eq") {
+			filter = filter + " and "
+		}
+		v := fmt.Sprintf("name eq '%s'", resourceName)
+		filter = filter + v
+	}
+
+	if resourceType != "" {
+		if strings.Contains(filter, "eq") {
+			filter = filter + " and "
+		}
+		v := fmt.Sprintf("resourceType eq '%s'", resourceType)
+		filter = filter + v
+	}
+
+	requiredTags := d.Get("required_tags").(map[string]interface{})
+
+	resources := make([]map[string]interface{}, 0)
+
+	resource, err := client.ListComplete(ctx, filter, "", nil)
+	if err != nil {
+		return fmt.Errorf("Error getting resources: %+v", err)
+	}
+
+	for resource.NotDone() {
+		res := resource.Value()
+
+		// currently the Azure-Go-SDK method "GetByID" does not work for some resources, as the
+		// API Version is hard coded, therefore we use ListComplete and look for the ResourceID 'manually'
+		if resourceID != "" && *res.ID != resourceID {
+			err = resource.NextWithContext(ctx)
+			if err != nil {
+				return fmt.Errorf("Error loading Resource List: %s", err)
+			}
+			continue
+		}
+
+		// currently its not supported to use a other filters together with the tags filter
+		// therefore we need to filter the resources manually.
+		tagMatches := 0
+		for requiredTagName, requiredTagVal := range requiredTags {
+			for tagName, tagVal := range res.Tags {
+				if requiredTagName == tagName && requiredTagVal == *tagVal {
+					tagMatches++
+				}
+			}
+		}
+
+		if tagMatches == len(requiredTags) {
+			s := make(map[string]interface{})
+
+			if v := *res.Name; v != "" {
+				s["name"] = v
+			}
```

Comment:
as of Terraform 0.12 this'll mean there's a nil value in the state rather than an empty string; could we change this to:

```suggestion
			name := ""
			if res.Name != nil {
				name = *res.Name
			}
```